spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarAdministration.java

```diff
@@ -0,0 +1,156 @@
+	}
+
+	private PulsarAdmin createAdminClient() throws PulsarClientException {
+		return adminBuilder.build();
+	}
+
+	@Override
+	public void createOrModifyTopics(PulsarTopic... topics) {
+		createOrModifyTopicsIfNeeded(Arrays.asList(topics));
+	}
+
+	private Map<String, List<PulsarTopic>> getTopicsPerNamespace(Collection<PulsarTopic> topics) {
+		return topics.stream().collect(Collectors.groupingBy(this::getTopicNamespaceIdentifier));
+	}
+	private String getTopicNamespaceIdentifier(PulsarTopic topic) {
+		return topic.getComponents().tenant() + "/" + topic.getComponents().namespace();
+	}
+
+	private List<String> getMatchingTopicPartitions(PulsarTopic topic, List<String> existingTopics) {
+		return existingTopics
+				.stream()
+				.filter(existing -> existing.startsWith(topic.getFullyQualifiedTopicName() + "-partition-"))
+				.toList();
+	}
+
+	private void createOrModifyTopicsIfNeeded(Collection<PulsarTopic> topics) {
+		if (CollectionUtils.isEmpty(topics)) {
+			return;
+		}
+
+		PulsarAdmin admin;
+		try {
+			admin = createAdminClient();
+		} catch (Exception e) {
+			throw new IllegalStateException("Could not create PulsarAdmin", e);
+		}
+
+		Map<String, List<PulsarTopic>> topicsPerNamespace = getTopicsPerNamespace(topics);
+
+		Set<PulsarTopic> topicsToCreate = new HashSet<>();
+		Set<PulsarTopic> topicsToModify = new HashSet<>();
+
+		topicsPerNamespace.forEach((namespace, requestedTopics) -> {
+			try {
+				List<String> existingTopicsInNamespace = admin.topics().getList(namespace);
+
+				for (PulsarTopic topic : requestedTopics) {
+					if (topic.isPartitioned()) {
+						List<String> matchingPartitions = getMatchingTopicPartitions(topic, existingTopicsInNamespace);
+						if (matchingPartitions.isEmpty()) {
+							logger.debug(() -> "Topic " + topic.getFullyQualifiedTopicName() + " does not exist.");
+							topicsToCreate.add(topic);
+						} else {
+							int numberOfExistingPartitions = matchingPartitions.size();
+							if (numberOfExistingPartitions < topic.numberOfPartitions()) {
+								logger.debug(() -> "Topic " + topic.getFullyQualifiedTopicName() + " found with "
+										+ numberOfExistingPartitions + " partitions.");
+								topicsToModify.add(topic);
+							} else if (numberOfExistingPartitions > topic.numberOfPartitions()) {
+								throw new IllegalStateException(
+										"Topic " + topic.getFullyQualifiedTopicName() + " found with "
+												+ numberOfExistingPartitions
+												+ " partitions. Needs to be deleted first.");
+							}
+						}
+					} else {
+						if (!existingTopicsInNamespace.contains(topic.getFullyQualifiedTopicName())) {
+							logger.debug(() -> "Topic " + topic.getFullyQualifiedTopicName() + " does not exist.");
+							topicsToCreate.add(topic);
+						}
+					}
+				}
+
+				createTopics(admin, topicsToCreate);
+				modifyTopics(admin, topicsToModify);
+			} catch (PulsarAdminException e) {
+				throw new RuntimeException(e);
+			}
+		});
+		admin.close();
```

Comment:
Not the end of the world probably, but this is will not be closed when topic creation fails above. The downside is that if we put this whole block in a giant try/resources or try/finally is that it way indented and less readable. One technique I do here is to move the logic to a `doXYZ` and then wrap it with a try/resources such as: 

```java
try (PulsarAdmin admin = createAdminClient()) {
	doCeateOrModifyTopicsIfNeeded(topics);
} catch (Exception e) {
	throw new IllegalStateException("Could not create PulsarAdmin", e);
}
```

Whatever the approach, we should go ahead and make sure this is closed eventhough it is an exceptional condition on startup only.